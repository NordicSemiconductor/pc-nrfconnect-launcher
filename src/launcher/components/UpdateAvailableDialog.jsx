/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';
import { bool, func, string } from 'prop-types';

/**
 * Dialog that is shown if an nRF Connect core update is available. The user
 * can either upgrade or cancel.
 *
 * @param {boolean} isVisible Show the dialog or not.
 * @param {string} version The new version number that is available.
 * @param {function} onClickReleaseNotes Invoked when the user clicks to see release notes.
 * @param {function} onConfirm Invoked when the user confirms the upgrade.
 * @param {function} onCancel Invoked when the user cancels the upgrade.
 * @returns {*} React element to be rendered.
 */
const UpdateAvailableDialog = ({
    isVisible,
    version,
    onClickReleaseNotes,
    onConfirm,
    onCancel,
}) => (
    <ConfirmationDialog
        isVisible={isVisible}
        title="Update available"
        text={
            `A new version (${version}) of nRF Connect is available. ` +
            'Would you like to upgrade now?'
        }
        okButtonText="Yes"
        cancelButtonText="No"
        onOk={onConfirm}
        onCancel={onCancel}
    >
        <p>
            A new version ({version}) of nRF Connect is available. Would you
            like to upgrade now?
        </p>
        <button
            className="btn btn-link core-btn-link"
            onClick={onClickReleaseNotes}
            type="button"
        >
            Click to see release notes
        </button>
    </ConfirmationDialog>
);

UpdateAvailableDialog.propTypes = {
    isVisible: bool.isRequired,
    version: string.isRequired,
    onClickReleaseNotes: func.isRequired,
    onConfirm: func.isRequired,
    onCancel: func.isRequired,
};

export default UpdateAvailableDialog;
