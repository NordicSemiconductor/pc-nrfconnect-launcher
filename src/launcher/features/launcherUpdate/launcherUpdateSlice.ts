/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import type { RootState } from '../../store';

const isWindows = process.platform === 'win32';
const isMac = process.platform === 'darwin';

export type State = {
    isUpdateAvailableDialogVisible: boolean;
    isUpdateProgressDialogVisible: boolean;
    isProgressSupported: boolean;
    isCancelSupported: boolean;
    latestVersion: string;
    percentDownloaded: number;
    isCancelling: boolean;
};

const initialState: State = {
    isUpdateAvailableDialogVisible: false,
    isUpdateProgressDialogVisible: false,
    isProgressSupported: false,
    isCancelSupported: false,
    latestVersion: '',
    percentDownloaded: 0,
    isCancelling: false,
};

const slice = createSlice({
    name: 'launcherUpdate',
    initialState,
    reducers: {
        updateAvailable(
            state,
            { payload: availableVersion }: PayloadAction<string>
        ) {
            state.isUpdateAvailableDialogVisible = true;
            state.latestVersion = availableVersion;
        },
        startDownload(
            state,
            {
                payload: {
                    isCancelSupported = isWindows,
                    isProgressSupported = isWindows || isMac,
                } = {},
            }: PayloadAction<
                | {
                      isCancelSupported?: boolean;
                      isProgressSupported?: boolean;
                  }
                | undefined
            >
        ) {
            state.isUpdateAvailableDialogVisible = false;
            state.isUpdateProgressDialogVisible = true;
            state.isProgressSupported = isProgressSupported;
            state.isCancelSupported = isCancelSupported;
        },
        cancelDownload(state) {
            state.isCancelling = true;
        },
        updateDownloading(state, action: PayloadAction<number>) {
            state.percentDownloaded = Math.floor(action.payload);
        },
        reset() {
            return initialState;
        },
    },
});

export default slice.reducer;

export const {
    updateAvailable,
    startDownload,
    cancelDownload,
    updateDownloading,
    reset,
} = slice.actions;

export const getLauncherUpdate = (state: RootState) => state.launcherUpdate;
