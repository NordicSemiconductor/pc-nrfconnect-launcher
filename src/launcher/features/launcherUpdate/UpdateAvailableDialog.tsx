/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import {
    ConfirmationDialog,
    ExternalLink,
} from '@nordicsemiconductor/pc-nrfconnect-shared';

import { inMain } from '../../../ipc/launcherUpdate';
import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { getLauncherUpdate, reset } from './launcherUpdateSlice';

const releaseNotesUrl =
    'https://github.com/NordicSemiconductor/pc-nrfconnect-launcher/releases';

export default () => {
    const dispatch = useLauncherDispatch();
    const {
        isUpdateAvailableDialogVisible: isVisible,
        latestVersion: version,
    } = useLauncherSelector(getLauncherUpdate);

    return (
        <ConfirmationDialog
            isVisible={isVisible}
            title="Update available"
            confirmLabel="Yes"
            cancelLabel="No"
            onConfirm={() => inMain.startUpdate()}
            onCancel={() => dispatch(reset())}
        >
            <p>
                A new version ({version}) of nRF Connect for Desktop is
                available. Would you like to update now?
            </p>
            <div className="tw-text-sm">
                <ExternalLink
                    label="Click to see release notes"
                    href={releaseNotesUrl}
                />
            </div>
        </ConfirmationDialog>
    );
};
