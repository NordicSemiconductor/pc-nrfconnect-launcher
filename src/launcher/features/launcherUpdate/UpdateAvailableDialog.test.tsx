/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import render from '../../../test/testrenderer';
import { updateAvailable } from './launcherUpdateSlice';
import UpdateAvailableDialog from './UpdateAvailableDialog';

describe('UpdateAvailableDialog', () => {
    it('is initially invisible', () => {
        expect(render(<UpdateAvailableDialog />).baseElement).toMatchSnapshot();
    });

    it('shows the new version if one becomes available ', () => {
        expect(
            render(<UpdateAvailableDialog />, [updateAvailable('1.2.3')])
                .baseElement,
        ).toMatchSnapshot();
    });
});
