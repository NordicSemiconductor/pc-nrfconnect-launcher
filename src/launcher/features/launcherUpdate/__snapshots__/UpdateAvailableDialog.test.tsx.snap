// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAvailableDialog is initially invisible 1`] = `
<body>
  <div />
</body>
`;

exports[`UpdateAvailableDialog shows the new version if one becomes available  1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade dialog  modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="tw-flex tw-flex-row"
          >
            <b>
              Update available
            </b>
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            A new version (
            1.2.3
            ) of nRF Connect for Desktop is available. Would you like to update now?
          </p>
          <div
            class="tw-text-sm"
          >
            <a
              class="tw-preflight tw-text-nordicBlue hover:tw-underline"
              href="https://github.com/NordicSemiconductor/pc-nrfconnect-launcher/releases"
              rel="noreferrer noopener"
              target="_blank"
            >
              Click to see release notes
            </a>
          </div>
        </div>
        <div
          class="tw-preflight tw-flex tw-flex-row-reverse tw-justify-start tw-gap-2 tw-border-none tw-p-4 modal-footer"
        >
          <button
            class="tw-preflight tw-h-8 tw-px-4 tw-text-sm tw-bg-nordicBlue tw-text-white active:enabled:tw-bg-nordicBlue-700 "
            type="button"
          >
            Yes
          </button>
          <button
            class="tw-preflight tw-h-8 tw-px-4 tw-text-sm tw-border tw-border-gray-700 tw-bg-white tw-text-gray-700  active:enabled:tw-bg-gray-50 "
            type="button"
          >
            No
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
